mirrord/agent/src/cgroup.rs

```diff
@@ -0,0 +1,256 @@
+        ))
+    }
+}
+
+const CGROUP_V1_FREEZE_PATH: &str = "freezer.state";
+
+/// We mkdir, mount the cgroup, which mounts the root for some reason
+/// then find the cgroup path, join it and write freeze to it.
+impl CgroupFreeze for CgroupV1 {
+    #[tracing::instrument(level = "trace", ret, skip(self))]
+    fn pause(&self) -> Result<()> {
+        // Check if our cgroup is mounted, if not, mount it.
+        let cgroup_path = Path::new(CGROUP_MOUNT_PATH);
+        if !cgroup_path.exists() {
+            trace!("mounting cgroup");
+            std::fs::create_dir(cgroup_path).map_err(|_| {
+                AgentError::PauseFailedCgroup("create dir failed cgroupv1".to_string())
+            })?;
+            mount(
+                Some("cgroup"),
+                cgroup_path,
+                Some("cgroup"),
+                MsFlags::MS_NOSUID | MsFlags::MS_NOEXEC | MsFlags::MS_NODEV,
+                Some("freezer"),
+            )
+            .map_err(|_| AgentError::PauseFailedCgroup("mount failed cgroupv1".to_string()))?;
+        }
+
+        let mut open_options = OpenOptions::new();
+        let mut file = open_options
+            .write(true)
+            .open(self.cgroup_path.join(CGROUP_V1_FREEZE_PATH))
+            .map_err(|_| AgentError::PauseFailedCgroup("open file cgroupv1 failed".to_string()))?;
+        file.write_all("FROZEN".as_bytes())
+            .map_err(|_| AgentError::PauseFailedCgroup("writing frozen failed".to_string()))?;
+        Ok(())
+    }
+
+    #[tracing::instrument(level = "trace", ret, skip(self))]
+    fn unpause(&self) -> Result<()> {
+        // if we're unpausing, mount should exist and we should be in the cgroup namespace
+        let mut open_options = OpenOptions::new();
+        let mut file = open_options
+            .write(true)
+            .open(self.cgroup_path.join(CGROUP_V1_FREEZE_PATH))?;
+        file.write_all("THAWED".as_bytes())?;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct CgroupV2 {}
+
+/// Reads given path's "cgroup.procs" file and returns the pids in it
+fn read_pids_cgroupv2(cgroup_path: &Path) -> Vec<u64> {
+    let file_name = cgroup_path.join(CGROUPV2_PROCS_FILE);
+    std::fs::File::open(file_name)
+        .map(|file| {
+            let bf = BufReader::new(file);
+            let mut v = Vec::new();
+            for line in bf.lines() {
+                match line {
+                    Ok(line) => match line.trim().parse() {
+                        Ok(n) => v.push(n),
+                        Err(_) => {
+                            warn!("Failed to parse pid {line}")
+                        }
+                    },
+                    Err(_) => break,
+                }
+            }
+            v
+        })
+        .unwrap_or_default()
+}
+
+/// Write given pids to given path's "cgroup.procs" file
+#[tracing::instrument(level = "trace", ret)]
+fn move_pids_to_cgroupv2(cgroup_path: &Path, pids: Vec<u64>) -> Result<()> {
```

Comment:
`cgroup_path` is being passed in some functions, why not put it into the `CgroupV2` struct, then turn these free functions into methods taking `&self`?

Reply:
It's being called twice with different path, so not really helpful